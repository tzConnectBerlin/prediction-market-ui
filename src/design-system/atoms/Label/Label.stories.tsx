import * as React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { GiAlarmClock } from 'react-icons/gi';
import { Label, LabelProps } from './Label';
import { lightTheme as theme } from '../../../styles/theme';

export default {
  title: 'Atom/Label',
  component: Label,
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as Meta;

const Template: Story<LabelProps> = (args) => <Label {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  text: 'Market',
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  text: 'Closed',
  backgroundColor: theme.palette.grey[500],
  fontColor: theme.palette.text.primary,
  icon: <GiAlarmClock />,
};

export const Small = Template.bind({});
Small.args = {
  size: 'small',
  text: 'Label',
};

export const Medium = Template.bind({});
Medium.args = {
  size: 'medium',
  text: 'Label',
  backgroundColor: theme.palette.warning.main,
};

export const Large = Template.bind({});
Large.args = {
  size: 'large',
  text: 'Label',
};
