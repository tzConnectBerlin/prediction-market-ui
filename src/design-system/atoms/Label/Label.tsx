import React from 'react';
import { SimplePaletteColorOptions, PaletteOptions } from '@material-ui/core/styles';
import styled from '@emotion/styled';
import { theme } from '../../../theme';

type PaletteOptionType = keyof PaletteOptions;
type PaletteColorOptionType = keyof SimplePaletteColorOptions;

interface StyledLabelProps {
  fontSize?: string;
  fontColor: string;
  backgroundColor?: string;
  iconSize?: string;
}

const StyledLabel = styled.span<StyledLabelProps>`
  font-size: ${({ fontSize }) => fontSize};
  border-radius: 0.2em;
  padding: 0.3em 0.6em;
  display: inline-block;
  background-color: ${({ backgroundColor }) => backgroundColor};
  color: ${({ fontColor }) => fontColor};
  &.label--icon {
    display: inline-flex;
    flex-direction: row;
    align-items: center;
    & > * {
      margin-right: 0.3em;
      font-size: ${({ iconSize }) => iconSize};
    }
  }
`;

export interface LabelProps {
  fontVariant?: PaletteOptionType;
  backgroundVariant?: PaletteOptionType;
  fontColor?: PaletteColorOptionType;
  backgroundColor?: PaletteColorOptionType;
  /**
   * Label text
   */
  text: string;
  /**
   * How large should the label be?
   */
  size?: 'small' | 'medium' | 'large';

  icon?: string | React.ReactNode;
}

/**
 * Primary UI component for user interaction
 */
export const Label: React.FC<LabelProps> = ({
  backgroundVariant = 'primary',
  backgroundColor = 'dark',
  text,
  size = 'small',
  icon,
  ...props
}) => {
  const hasIcon = icon ? 'label--icon' : '';
  const fontSize =
    size && (size.includes('em') || size.includes('px'))
      ? size
      : size === 'large'
      ? '1.3em'
      : size === 'medium'
      ? '1em'
      : '0.8em';
  const iconSize = hasIcon && size && size === 'small' ? '1.1em' : fontSize;

  /**
   * TODO: find a way to do this without `any`
   */
  const internalBackgroundVariant: any = theme.palette[backgroundVariant];
  const internalBackgroundColor = internalBackgroundVariant[backgroundColor];

  const fontColor: string =
    backgroundColor && backgroundColor === 'dark'
      ? theme.palette.lightText.primary
      : theme.palette.text.primary;

  return (
    <StyledLabel
      fontSize={fontSize}
      fontColor={fontColor}
      backgroundColor={internalBackgroundColor}
      iconSize={iconSize}
      className={hasIcon}
      {...props}
    >
      {icon}
      {text}
    </StyledLabel>
  );
};
