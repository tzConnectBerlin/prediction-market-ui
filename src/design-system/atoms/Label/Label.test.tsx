import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { GiAlarmClock } from 'react-icons/gi';
import { Label } from './Label';

describe('Snapshot testing Label Component', () => {
  it('renders correctly with small size and primary color', () => {
    const label = renderer.create(<Label text="Label Text" />).toJSON();
    expect(label).toMatchSnapshot();
  });

  it('renders correctly with medium size and secondary background color', () => {
    const label = renderer
      .create(
        <Label
          text="Label Text"
          backgroundVariant="secondary"
          backgroundColor="light"
          size="medium"
        />,
      )
      .toJSON();
    expect(label).toMatchSnapshot();
  });

  it('renders correctly with large size', () => {
    const label = renderer.create(<Label text="Label Text" size="large" />).toJSON();
    expect(label).toMatchSnapshot();
  });

  it('renders correctly with icon', () => {
    const label = renderer.create(<Label text="Label Text" icon={<GiAlarmClock />} />).toJSON();
    expect(label).toMatchSnapshot();
  });
});

describe('Element testing Label Component', () => {
  it('render correctly with icon and label--icon class', () => {
    const { container } = render(<Label text="Label Text" icon={<GiAlarmClock />} />);
    expect(container.firstChild).toHaveClass('label--icon');
  });

  it('render correctly with Market text', () => {
    const { container } = render(<Label text="Market" />);
    expect(container.textContent).toBe('Market');
  });
});
