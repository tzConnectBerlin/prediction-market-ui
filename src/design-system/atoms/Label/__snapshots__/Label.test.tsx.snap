// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing Label Component renders correctly with icon 1`] = `
.emotion-0 {
  border-radius: 0.2em;
  padding: 0.3em 0.6em;
  display: inline-block;
  background-color: #1976d2;
  color: #fff;
}

.emotion-0.label--icon {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0.label--icon>* {
  margin-right: 0.3em;
}

.emotion-0.label--icon>svg {
  width: 0.9em;
  height: 0.9em;
}

.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
}

.emotion-1.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

<span
  className="label--icon emotion-0"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M247 62.107v9.94h-28.88v18H247v39.074c-84.367 4.682-151.514 74.737-151.514 160.26 0 41.037 15.468 78.508 40.873 106.91l-36.237 36.237 12.727 12.73 36.238-36.24C177.49 434.425 214.96 449.894 256 449.894c41.04 0 78.51-15.47 106.912-40.875l36.238 36.238 12.727-12.73-36.236-36.235c25.406-28.402 40.874-65.873 40.874-106.91 0-85.523-67.147-155.578-151.514-160.26V90.048h28.88v-18H265v-9.94h-18zm-104.906 26.58c-3.583-.03-7.267.125-11.014.465C111.1 90.97 89.15 98 73.637 113.514c-15.514 15.513-22.545 37.462-24.362 57.443-.908 9.99-.488 19.54 1.184 27.898 1.67 8.36 4.315 15.662 9.784 21.13l6.363 6.365 6.366-6.364 113.5-113.502-6.366-6.363c-5.47-5.468-12.77-8.112-21.128-9.784-4.18-.836-8.658-1.358-13.34-1.56-1.17-.05-2.352-.08-3.546-.088zm227.812 0c-1.194.01-2.376.04-3.547.09-4.682.2-9.16.723-13.34 1.56-8.357 1.67-15.658 4.315-21.127 9.784l-6.366 6.364 6.366 6.366 113.5 113.5 6.363-6.364c5.47-5.47 8.113-12.772 9.785-21.13 1.673-8.36 2.093-17.908 1.185-27.9-1.817-19.98-8.848-41.93-24.362-57.442C422.85 98 400.9 90.97 380.92 89.152c-3.747-.34-7.43-.494-11.014-.465zM256 146.865c78.815 0 142.514 63.7 142.514 142.514 0 78.814-63.7 142.513-142.514 142.513-78.815 0-142.514-63.7-142.514-142.514 0-78.816 63.7-142.515 142.514-142.515zm-88.332 41.453l-12.73 12.73 82.92 82.917a18.94 18.94 0 0 0-.797 5.414A18.94 18.94 0 0 0 256 308.317a18.94 18.94 0 0 0 16.66-9.94h59.096v-18h-59.11A18.94 18.94 0 0 0 256 270.44a18.94 18.94 0 0 0-5.408.804l-82.924-82.924z"
    />
  </svg>
  <div
    className="MuiTypography-root MuiTypography-h5  emotion-1"
    lines={3}
  >
    Label Text
  </div>
</span>
`;

exports[`Snapshot testing Label Component renders correctly with large size 1`] = `
.emotion-0 {
  border-radius: 0.2em;
  padding: 0.3em 0.6em;
  display: inline-block;
  background-color: #1976d2;
  color: #fff;
}

.emotion-0.label--icon {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0.label--icon>* {
  margin-right: 0.3em;
}

.emotion-0.label--icon>svg {
  width: 1.3em;
  height: 1.3em;
}

.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
}

.emotion-1.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

<span
  className=" emotion-0"
>
  <div
    className="MuiTypography-root MuiTypography-h5  emotion-1"
    lines={3}
  >
    Label Text
  </div>
</span>
`;

exports[`Snapshot testing Label Component renders correctly with medium size and secondary background color 1`] = `
.emotion-0 {
  border-radius: 0.2em;
  padding: 0.3em 0.6em;
  display: inline-block;
  background-color: rgba(1, 102, 255, 0.12);
  color: rgba(29, 34, 39, 0.87);
}

.emotion-0.label--icon {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0.label--icon>* {
  margin-right: 0.3em;
}

.emotion-0.label--icon>svg {
  width: 1em;
  height: 1em;
}

.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
}

.emotion-1.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

<span
  className=" emotion-0"
>
  <div
    className="MuiTypography-root MuiTypography-h5  emotion-1"
    lines={3}
  >
    Label Text
  </div>
</span>
`;

exports[`Snapshot testing Label Component renders correctly with small size and primary color 1`] = `
.emotion-0 {
  border-radius: 0.2em;
  padding: 0.3em 0.6em;
  display: inline-block;
  background-color: #1976d2;
  color: #fff;
}

.emotion-0.label--icon {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0.label--icon>* {
  margin-right: 0.3em;
}

.emotion-0.label--icon>svg {
  width: 0.8em;
  height: 0.8em;
}

.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
}

.emotion-1.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

<span
  className=" emotion-0"
>
  <div
    className="MuiTypography-root MuiTypography-h5  emotion-1"
    lines={3}
  >
    Label Text
  </div>
</span>
`;
